Zend/zend_vm_def.h

```diff
@@ -8402,6 +8402,69 @@ ZEND_VM_COLD_CONSTCONST_HANDLER(188, ZEND_SWITCH_STRING, CONST|TMPVARCV, CONST,
 	}
 }
 
+ZEND_VM_COLD_CONSTCONST_HANDLER(195, ZEND_MATCH_LONG, CONST|TMPVARCV, CONST, JMP_ADDR)
+{
+	USE_OPLINE
+	zval *op, *jump_zv;
+	HashTable *jumptable;
+
+	op = GET_OP1_ZVAL_PTR_UNDEF(BP_VAR_R);
+
+	if (Z_TYPE_P(op) != IS_LONG) {
+		ZVAL_DEREF(op);
+		if (Z_TYPE_P(op) != IS_LONG) {
```

Comment:
So, there's a minor edge case I overlooked.

I think it works in regular ZEND_SWITCH_LONG and ZEND_SWITCH_STRING because it falls through to ZEND_CASE and ZEND_IS_EQUAL opcodes, which do check if CV (i.e. direct references to $varName) are undefined. (so that doesn't need changing)

https://github.com/php/php-src/pull/4914/files is one example of how to properly check an operand being undefined, and a unit test for a similar bug I've seen in the past.

- If an exception is thrown or a notice is emitted fetching the undefined variable, we want to have the correct stack trace (hence the SAVE_OPLINE), and if the exception is thrown, we want to go to the new opline for the exception instead of the default branch.

```suggestion
		if (Z_TYPE_P(op) != IS_LONG) {
			if (UNEXPECTED((OP1_TYPE & IS_CV) && Z_TYPE_P(op) == IS_UNDEF)) {
				SAVE_OPLINE();
				ZVAL_UNDEFINED_OP1();
				if (UNEXPECTED(EG(exception))) {
					HANDLE_EXCEPTION();
				}
			}
```

A potential test to validate this would be fixed (I haven't ran this yet):

```php
set_error_handler(function ($errno, $message) {
    throw new Exception("Custom error handler: $message");
});
match ($undefVar) {
   default => print("This should not get printed with or without opcache\n"),
   1, 2, 3, 4, 5 => print("Also should not be printed"),
}
print("unreachable");
```

Reply:
Totally, your example fails. I would've probably never caught that. Thanks!

Reply:
Another test case to add

Observed: Emits UnhandledMatchError
Expected: Should emit an undefined variable notice for $undef, then emit UnhandledMatchError

```
» php -a
Interactive shell

php > match($undef) {}

Warning: Uncaught UnhandledMatchError in php shell code:1
Stack trace:
#0 {main}
  thrown in php shell code on line 1
php > echo $undef2;

Warning: Undefined variable $undef2 in php shell code on line 1
```

In order to fix that, it seems like CHECK_VAR should be emitted if no IS_IDENTICAL opcodes would be emitted (i.e. `match($x) {}` or `match($x) { default => '...' }`.

Interestingly enough, I discovered that `switch($x) {}` has the exact same bug for not warning about the undefined variable $x (I'm assuming this is a bug, anyway).

```
» sapi/cli/php -d zend_extension=$PWD/modules/opcache.so -d opcache.enable_cli=1 -d opcache.enable=1 -d opcache.opt_debug_level=0x30000 test.php

$_main:
     ; (lines=5, args=0, vars=1, tmps=3)
     ; (before optimizer)
     ; /path/to/test.php:1-3
     ; return  [] RANGE[0..0]
0000 JMP 0001
0001 V2 = NEW 0 string("UnhandledMatchError")
0002 DO_FCALL
0003 THROW V2
0004 RETURN int(1)
LIVE RANGES:
     2: 0002 - 0003 (new)

$_main:
     ; (lines=3, args=0, vars=0, tmps=1)
     ; (after optimizer)
     ; /path/to/test.php:1-3
0000 V0 = NEW 0 string("UnhandledMatchError")
0001 DO_FCALL
0002 THROW V0
LIVE RANGES:
     0: 0001 - 0002 (new)

Fatal error: Uncaught UnhandledMatchError in /usr/local/tyson/php-src/test.php:2
Stack trace:
#0 {main}
  thrown in /usr/local/tyson/php-src/test.php on line 2
```

Reply:
Interestingly, this runs without errors:

```php
switch ($undefVar) {}
```

Probably because the value is never checked against anything.